File: Client/src/Packets/ACK.cpp
//
// Created by dorgreen on 1/17/17.
//

#include "include/Packets/ACK.h"
#include "Packet.cpp"


using namespace std;

// @ Sends these packets to server ==> constructor from data
// @ Receives these packets ==> constructor from char*
ACK::ACK(char *incoming) : Packet(Opcode::_ACK) {
    block_num = Packet::bytesToShort(incoming); // TODO: if does include the opcode in the char*, add logic here
}

ACK::ACK(short block_num) : Packet(Opcode::_ACK), block_num(block_num) {}



char *ACK::toBytes() {

    char opcode_bytes[2];
    shortToBytes(Opcode::_ACK, opcode_bytes);
    char block_num_bytes[2];
    shortToBytes(block_num, block_num_bytes);

    char *toReturn = new char[4]{opcode_bytes[0], opcode_bytes[1], block_num_bytes[0], block_num_bytes[1]};


    return toReturn;

}

ACK::~ACK() {

}

int ACK::getBytesCount() {
    return 4;
}

File: Server/src/main/java/bgu/spl171/net/srv/ACK.java
package bgu.spl171.net.srv;

import java.util.Arrays;

/**
 * Created by dorgreen on 20/01/2017.
 */
public class ACK extends Packet {

    protected int block_number;

    protected ACK() {
        super((short) 4);
        block_number = ACK_SUCCESSFUL; // Defined in the protocol as the default value for anything but data packets
    }

    protected ACK(int block_number) {
        super((short) 4);
        this.block_number = block_number;
    }

    @Override
    public Packet getPacket(byte[] bytes, int numOfBytes) {
        if (numOfBytes == 4) {
            this.block_number = bytesToShort(Arrays.copyOfRange(bytes, 2, 4));
            return this;
        } else return null;
    }

    // TODO: IMPLEMENT
    @Override
    public Packet handle(MessagingProtocolImpl protocol) {
        // This should communicate with packets awaiting acknowledgment

        // Send the next DATA packet awaiting confirmation, whether it's RRQ or DIRQ
        if (block_number > 0) {
            protocol.sendNextDataPacket();
        }

        return null;
    }

    @Override
    public byte[] toBytes() {
        byte[] ans = new byte[OP_CODE_SIZE + 2];
        insertArrayIntoArray(ans, shortToBytes(OP_CODE), 0);
        insertArrayIntoArray(ans, shortToBytes((short) block_number), 2);
        return ans;
    }
}
